### PR TITLE
[Java] Document fixes for deserialization vulnerabilities by framework

### PR DESCRIPTION
**Pull request by JLLeitschuh**
Related https://github.com/github/codeql/issues/11603

### CODE REVIEW

File: java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp [hunk 1]
```diff
@@ -51,6 +51,56 @@ from the input stream removes the vulnerability.
...
+            <td>com.fasterxml.jackson.core:jackson-databind</td>
+            <td>Yes</td>
+            <td>Don't call <code>com.fasterxml.jackson.databind.ObjectMapper#enableDefaultTyping</code>. Read <a href="https://cowtowncoder.medium.com/jackson-2-10-safe-default-typing-2d018f0ce2ba">this guide</a>.</td>
+        </tr>
+        <tr>
+            <td>Kryo</td>
+            <td>com.esotericsoftware:kryo and com.esotericsoftware:kryo5</td>
+            <td>com.esotericsoftware:kryo versions after 5.0 Yes; com.esotericsoftware:kryo5 Yes</td>
+            <td>Don't call <code>com.esotericsoftware.kryo(5).Kryo#setRegistrationRequired</code> with the argument <code>false</code>.</td>
+        </tr>
```
Comment (JLLeitschuh) [inline java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp:100]: I did some research on this and Kryo is no longer vulnerable by default. It seems that sometime between the end of the 4.x release line and the latest release this was fixed to be secure by default.

Currently, the Kryo query still present in CodeQL checks for this incorrectly.

CodeQL currently attempts to detect calls to `setRegistrationRequired(true)` for the use case to be safe. But actually, the library currently requires the user to call `setRegistrationRequired(false)` to be vulnerable. The logic in CodeQL needs to be inverted to be correct.

File: java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp [hunk 2]
```diff
@@ -51,6 +51,56 @@ from the input stream removes the vulnerability.
...
+            <td>com.alibaba:fastjson</td>
+            <td>Partially</td>
+            <td>Call <code>com.alibaba.fastjson.parser.ParserConfig#setSafeMode</code> with the argument <code>true</code></td>
+        </tr>
+        <tr>
+            <td>SnakeYAML</td>
+            <td>org.yaml:snakeyaml</td>
+            <td><a href="https://bitbucket.org/snakeyaml/snakeyaml/wiki/CVE%20&amp;%20NIST.md">No</a>. <a href="https://bitbucket.org/snakeyaml/snakeyaml/issues/561/cve-2022-1471-vulnerability-in">Maintainer response</a>.</td>
+            <td>Instantiate the <code>org.yaml.snakeyaml.Yaml</code> instance explicitly with an instance of <code>org.yaml.snakeyaml.constructor.SafeConstructor</code> as an argument.</td>
+        </tr>
```
Comment (JLLeitschuh) [inline java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp:88]: This has been my pet project the past two weeks. I'm still working to convince the maintainer to make this secure by default. If this does end up happening, I'll update this documentation.

File: java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp (hunk 1 above)
Comment (JLLeitschuh) [reply java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp:100]: I'm trying to track down the exact version this was fixed. I've reached out to the maintainers here: https://github.com/EsotericSoftware/kryo/issues/929

Comment (JLLeitschuh) [reply java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp]: Confirmed that versions 5.0.0 and later are all secure-by-default.

**Comment by smowton** [comment]
@JLLeitschuh we would rather not have specific research re: when and how particular libs are vulnerable in qhelp files -- it would be better to get documentation such as this published by a security research organisation (e.g. integrated into https://cheatsheetseries.owasp.org/cheatsheets/Deserialization_Cheat_Sheet.html ?), then cite that from our qhelp.

**Comment by JLLeitschuh** [comment]
This seems like a contradiction. All of the QHelp files have bad and good examples. How is this any different?

This knowledge is currently heavily scattered across the internet and very difficult to track down. Also, the current CodeQL deserialization QHelp is resulting in incorrect understandings by end users. As an example, the Oracle team incorrectly understood the SnakeYaml vulnerability due to CodeQLs documentation.

https://www.websec.ca/publication/Blog/CVE-2022-21404-Another-story-of-developers-fixing-vulnerabilities-unknowingly-because-of-CodeQL

**Comment by smowton** [comment]
@JLLeitschuh I think the goal is that qhelp should contain a general example of the class of vulnerability and how it should be handled, but not a library-by-library how-to guide, since the latter will get out of date and is better hosted by an organisation that's particularly in that business, e.g. OWASP. I think you should get it hosted by a reputable authority and then link it from our qhelp using phrasing like "...for specific advice about how to secure a particular library, see..."

**Comment by JLLeitschuh** [comment]
> I think the goal is that qhelp should contain a general example of the class of vulnerability and how it should be handled, but not a library-by-library how-to guide, since the latter will get out of date and is better hosted by an organization that's particularly in that business, e.g. OWASP.

This is exactly what CodeQL already does. As QL query authors, you have to encode in the query what "safe" means, otherwise the query will issue false positives all over the place. The problem here, is that that knowledge that is encoded in the query, is not captured in the documentation associated with the query. This seems wrong.

As an end user, why am I getting this alert? If you have two uses of some chunk of code in your codebase, why is one getting flagged, and the other isn't? Most likely, this is because of intentional filtering encoded into the ql query. Even as someone who writes ql, it can be quite complicated to figure out why one chunk of code is/isn't being flagged.

> I think you should get it hosted by a reputable authority

GitHub has made itself this reputable authority by flagging code as vulnerable. It is the responsibility of the query authors to be able to explain, both in CodeQL, and in the documentation that query flags, why the vulnerability was flagged. CodeQL (either through the documentation or through the query results) should be able to communicate this per-library.

TL;DR: GitHub and CodeQL is already encoding this information in the query, even if it is a little out of date. If CodeQL has the responsibility of flagging a vulnerability, the documentation should be able to adequately describe _why_ it is flagging it as vulnerable.

**Comment by JLLeitschuh** [comment]
> the documentation should be able to adequately describe _why_ it is flagging it as vulnerable.

If you look at the queries, this is often because the user has/hasn't done something to make their use case vulnerable. That information about what they need to do/haven't done is essential to actually fixing the vulnerability. Again, CodeQL is capturing this in the query, but isn't articulating this to the end user.

To provide a concrete example:
https://github.com/github/codeql/blob/e629568eda11c9055d7d279b3a40b818e9c6b59e/java/ql/lib/semmle/code/java/security/UnsafeDeserializationQuery.qll#L133-L142

CodeQL has already encoded in the query that this is a safe way to do object deserialization, but the documentation doesn't reflect this.

Another example:
https://github.com/github/codeql/blob/e629568eda11c9055d7d279b3a40b818e9c6b59e/java/ql/lib/semmle/code/java/security/UnsafeDeserializationQuery.qll#L181-L189

CodeQl has included in the query a search for the code that makes Jackson vulnerable, but this also doesn't get captured anywhere in the documentation.

I get that it's difficult to keep documentation and code in sync. If that's truly the problem then, maybe the _why_ should be more explicitly communicated in the query results. But it should be somewhere.

Without this _why_ end users are far more likely to simply consider the result a false positive. They can't be expected to do extensive research on every vulnerability. As an example, I've spent the past week learning about deserialization vulnerabilities and the various gadget chains that exist.

If I were able to wave a magic wand, I would completely re-write this query so that every framework is it's own query result and each framework is documented independently. As an end-user this would be the most clear because the documentation would be more narrowly focused on my particular use case.

**Comment by atorralba** [comment]
I made some minor changes so that the query help file renders correctly, I hope you don't mind @JLLeitschuh. That should help with the review to determine whether to accept this.

**Comment by coadaflorin** [comment]
@JLLeitschuh thanks for gathering this data and creating this PR. The team and I discussed and we agree that this is helpful information for developers to have when trying to understand what the problem is and how to fix it. At the moment the way in which the table is formatted is creating horizontal scrolling and hiding some essential information from the user. Is it possible to try a format that requires less space horizontally?

We'll also plan some adjustments to avoid showing findings for libraries which are secure by default that developers will report as False Positives.

**Comment by JLLeitschuh** [comment]
Hi All,

I'm on vaacation until the 28th so won't have a response/fix until then. Do you have any suggestions for alternative formatting? I'm open to anything.

**Comment by coadaflorin** [comment]
There is a bot response that shows how the message would look like: 
https://github.com/github/codeql/pull/11700#issuecomment-1357325140 

Clicking the twisty should reveal the rendered response. I'll reach out to the internal documentation team and see if they have a suggestion for us.

File: java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp [hunk 3]
```diff
@@ -28,11 +28,64 @@ for example JSON or XML.  However, these formats should not be deserialized
...
 are possible through libraries such as XStream and XmlDecoder.
-
+</p>
+<p>
 Alternatively, a tightly controlled whitelist can limit the vulnerability of code, but be aware
 of the existence of so-called Bypass Gadgets, which can circumvent such
 protection measures.
 </p>
+<p>
+Fixes by framework:
```
Comment (smowton) [inline java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp:38]:
```suggestion
Recommendations specific to particular frameworks supported by this query:
```

(So, should we make sure we say something about every framework mentioned in the `<overview>` section, even if it's just to note that we don't look for any particular configuration and just alert on all uses with potentially-tainted input?)

File: java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp [hunk 4]
```diff
@@ -28,11 +28,64 @@ for example JSON or XML.  However, these formats should not be deserialized
...
+            <th>Project</th>
+            <th>Maven Coordinates</th>
+            <th>Secure by Default</th>
+            <th>Fix</th>
+        </tr>
+            <tr>
+            <td>XMLDecoder</td>
+            <td>Java Standard Library</td>
+            <td>No</td>
+            <td>Don't use XMLDecoder with untrusted user input. It is impossible to secure.</td>
```
Comment (smowton) [inline java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp:52]:
```suggestion
            <td>Do not use with untrusted user input.</td>
```

File: java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp [hunk 5]
```diff
@@ -28,11 +28,64 @@ for example JSON or XML.  However, these formats should not be deserialized
...
+            <td>org.yaml:snakeyaml</td>
+            <td><a href="https://bitbucket.org/snakeyaml/snakeyaml/wiki/CVE%20&amp;%20NIST.md">No</a> (<a href="https://bitbucket.org/snakeyaml/snakeyaml/issues/561/cve-2022-1471-vulnerability-in">maintainer response</a>)</td>
+            <td>Instantiate the <code>org.yaml.snakeyaml.Yaml</code> instance explicitly with an instance of <code>org.yaml.snakeyaml.constructor.SafeConstructor</code> as an argument.</td>
+        </tr>
+        <tr>
+            <td>FasterXML jackson-databind</td>
+            <td>com.fasterxml.jackson.core:jackson-databind</td>
+            <td>Yes</td>
+            <td>
+                Don't call <code>com.fasterxml.jackson.databind.ObjectMapper#enableDefaultTyping</code> and don't annotate any object fields with <code>com.fasterxml.jackson.annotation.JsonTypeInfo</code> passing either the <code>CLASS</code> or <code>MINIMAL_CLASS</code> values to the annotation.
```
Comment (smowton) [inline java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp:77]:
```suggestion
                Don't call <code>com.fasterxml.jackson.databind.ObjectMapper#enableDefaultTyping</code> and don't annotate any object fields with <code>@JsonTypeInfo(CLASS) or @JsonTypeInfo(MINIMAL_CLASS)</code> if untrusted data may be deserialized.
```

**Review by smowton — COMMENTED** [review]
Some style suggestions, independent of the question of table formatting

File: java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp [hunk 6]
```diff
@@ -28,11 +28,64 @@ for example JSON or XML.  However, these formats should not be deserialized
...
+                Read <a href="https://cowtowncoder.medium.com/jackson-2-10-safe-default-typing-2d018f0ce2ba">this guide</a>.
```
[the remark this answers is not in the document]
Comment (JLLeitschuh) [reply java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.qhelp:78]: Cowtowncoder is the primary maintainer of FasterJackson

**Comment by JLLeitschuh** [comment]
Update: SnakeYaml is actually fixing the library so it's not insecure-by-default. 🎉 

https://bitbucket.org/snakeyaml/snakeyaml/issues/561/cve-2022-1471-vulnerability-in

**Comment by JLLeitschuh** [comment]
@coadaflorin the links in the table don't seem to be rendering. Is that an issue with the markdown generator?

**Comment by JLLeitschuh** [comment]
I think I'm inclined to go with the list format you proposed. Do you want to create a PR against my branch and I'll merge it into this PR?

**Comment by coadaflorin** [comment]
I think I managed to create the correct PR: https://github.com/JLLeitschuh/ql/pull/39. 
Hope that helps :)

**Comment by JLLeitschuh** [comment]
It looks like your PR has conflicts as-is. Can you resolve those? After that, I'm happy to merge it

**Comment by coadaflorin** [comment]
Looks like the list is in, I'm all good on my side.

**Review by atorralba — APPROVED** [review]

**Review by smowton — APPROVED** [review]
